Add logical selection to loc()

mreijnde opened the pull request:
Add logical selection to loc() and move index selection code to Index class

Restructured subsasign to use these new selector function in Index class. What is your view on this?

todo: update testcases + add test cases

mreijnde commented:
I think the functionality now works, including the testcases. Could you have a look? 

- Index class has now all code to make index selection and validations
    -  getSelector() output a matlab compatible indexer (colon, logical array, or position index array) 
    -  positionOf() outputs always an position index array

- Assignments of dataframe selection code rearranged to split into
    - 1d indexing cases using Index.getSelector() method
    - 2d indexing (logical matrix) cases handled by modifyFromBool2D()
    
- Other DataFrame methods, eg dropIndex(), dropColumns() that use the selection will automatically support logical indexing

benjamingaudin reviewed (changes requested):
Your code follows a good logic and I'll be happy to add your changes after you read my comments

mreijnde commented on +frames/DataFrame.m:
```diff
@@ -1046,55 +1046,71 @@
+                             error("No arguments given for .%s()", field);
+                         elseif length(s)>2
+                             error("Nested assign in combination with .%s() indexing not supported", field);
+                         end
```
indeed, makes code clearer

mreijnde commented on +frames/DataFrame.m:
```diff
@@ -1168,84 +1167,80 @@ function columnsValidation(obj,value)
+            assert( size(data,1) == idxN || size(data,1)<=1, 'frames:modify:WrongDataDimension', ...
+                "not matching number of rows in supplied data");
+            assert( size(data,2) == colN || size(data,2)<=1, 'frames:modify:WrongDataDimension', ...
+                "not matching number of cols in supplied data");            
```
These lines, including lines 1183-1189 are a size check  between data and selector, with idea to get custom error message instead of failing on assignment. 1197-1201 is size between existing index.

Removed the initial size check, rely on just the internal matlab check.

mreijnde commented on +frames/DataFrame.m:
```diff
@@ -1168,84 +1167,80 @@ function columnsValidation(obj,value)
+        function obj = modifyFromBool2D(obj, subs, b)
+            % modify selected data by 2D logical (matrix or logical dataframe) to supplied data
+            assert( length(subs)==1, 'frames:modifyFromBool2D:OnlySingleSelectorAllowed' , ...
+                 'Only single 2d logical selector allowed');
```
moved check to subsasgn

mreijnde commented on +frames/Index.m:
```diff
@@ -264,6 +330,50 @@ function valueChecker(obj,value,fromSubsAsgnIdx,b)
+                'frames:positionIndexChecker:InvalidPositionIndexType', ...
+                 "Only positive integers allowed as position index");
+            assert(all(selector<=length(obj)), 'frames:positionIndexChecker:PositionIndexOutOfRange', ...
+                "Position index out of range (%d)", length(obj));        
```
removed them

mreijnde commented on +frames/Index.m:
```diff
@@ -264,6 +330,50 @@ function valueChecker(obj,value,fromSubsAsgnIdx,b)
+            if isFrame(selector)
+                % convert logical series to logical array after checks
+                assert(islogical(selector.data_), 'frames:logicalIndexChecker:LogicalFrameSeriesRequired', ...
+                       "Only logical DataFrame-Series allowed for indexing");
```
k

mreijnde commented on +frames/Index.m:
```diff
@@ -264,6 +330,50 @@ function valueChecker(obj,value,fromSubsAsgnIdx,b)
+            %assert(length(selector)==length(obj.value_), 'frames:logicalIndexChecker:WrongDimension', ...
+            %       "Logical index array should have same size as dimension.", length(obj.value_));
+            assert(isvector(selector), 'frames:logicalIndexChecker:VectorRequired', ...
+                   "Logical index array should be vector (no matrix)");                            
```
k

mreijnde commented:
Thx for the detailed reviewing, the code improved. I have updated the pull request with the review comments, and added some additional testcases. Please let me know if something else has to change.

benjamingaudin commented:
Awesome work!
